### ISSUE TITLE
field with type int64 as primary key, cache no hit

### ISSUE THREAD

**Issue by yulefox** [go-xorm/xorm#220]
type DatRole {
        Id  uint64
        Name string
    }

    func GetRole(name string) []DataRole {
        roles :=make([]DatRole, 0)
        //engine.NoCache()
        engine.Where("name like ?", name).Find(&roles)
        return roles
    }

First request (no cache):

    [xorm] [debug] 2015/03/13 17:37:35.035771 [cacheFind] cache sql: [[999001423731429334] [999001423732209317] [999001423732250085] [999001424857132810]] dat_role SELECT `dat_role`.`id` FROM  `dat_role` WHERE user like ? [test_user]
    [xorm] [info]  2015/03/13 17:37:35.035866 [sql] SELECT `id`, `name` FROM `dat_role` WHERE (`id` IN (?,?,?,?)) [args] [999001423731429334 999001423732209317 999001423732250085 999001424857132810]
    [xorm] [debug] 2015/03/13 17:37:35.036251 [cacheFind] cache bean: dat_role [999001423731429334] &{999001423731429334 test_user}
    [xorm] [debug] 2015/03/13 17:37:35.036264 [cacheFind] cache bean: dat_role [999001423732209317] &{999001423732209317 test_user}
    [xorm] [debug] 2015/03/13 17:37:35.036273 [cacheFind] cache bean: dat_role [999001423732250085] &{999001423732250085 test_user}
    [xorm] [debug] 2015/03/13 17:37:35.036285 [cacheFind] cache bean: dat_role [999001424857132810] &{999001424857132810 test_user}

Second request (cache hit, however, numbers type of `int64` are converted into `NUMe+n`):

    [xorm] [debug] 2015/03/13 17:38:08.380231 [cacheFind] cache hit sql: SELECT `dat_role`.`id` FROM  `dat_role` WHERE user like ? [test_user]
    [xorm] [info]  2015/03/13 17:38:08.380278 [sql] SELECT `id`, `name` FROM `dat_role` WHERE (`id` IN (?,?,?,?)) [args] [9.990014237314294e+17 9.990014237322093e+17 9.990014237322501e+17 9.990014248571328e+17]
    [xorm] [warn]  2015/03/13 17:38:08.380486 [cacheFind] cache no hit: dat_role [9.990014237314294e+17]
    [xorm] [warn]  2015/03/13 17:38:08.380490 [cacheFind] cache no hit: dat_role [9.990014237322093e+17]
    [xorm] [warn]  2015/03/13 17:38:08.380492 [cacheFind] cache no hit: dat_role [9.990014237322501e+17]
    [xorm] [warn]  2015/03/13 17:38:08.380494 [cacheFind] cache no hit: dat_role [9.990014248571328e+17]

**Comment by lunny** [comment]
I think maybe you should add `pk` tag to `Id` for indicate it's a primary key column.

```Go
type DatRole {
    Id  uint64 `xorm:"pk"`
    Name string
}
```

**Comment by lunny** [comment]
This bug is caused by PK marshaled to json and unmarshaled from json.
When unmarshaled from json, encoding/json will convert 999001424857132810 to a float not an uint64.

**Comment by lunny** [comment]
fixed by github.com/go-xorm/xorm#be6e7ac47dc57bd0ada25322fa526944f66ccaa6

**Comment by yulefox** [comment]
well done, thank you.

2015-03-15 8:32 GMT+08:00 Lunny Xiao <notifications@github.com>:

> Closed #220 <https://github.com/go-xorm/xorm/issues/220>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/go-xorm/xorm/issues/220#event-254107999>.
>



-- 
_______________________________
   http://www.yulefox.com
